semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -102,26 +114,28 @@ parseFile path = do
     -- FIXME: this should use the span of the source file, not an empty span.
     Right (_, Just root) -> pure (File (Ref.fromPath (Path.absRel path)) root)
 
-parseGraph :: A.Value -> A.Parser (IntMap.IntMap Term, Maybe Term)
+newtype Graph = Graph { terms :: IntMap.IntMap Term }
+
+parseGraph :: A.Value -> A.Parser (Graph, Maybe Term)
 parseGraph = A.withArray "nodes" $ \ nodes -> do
-  (untied, First root) <- foldMap (\ (k, v, r) -> ([(k, v)], First r)) <$> traverse (A.withObject "node" parseNode) (V.toList nodes)
-  -- @untied@ is a list of key/value pairs, where the keys are graph node IDs and the values are functions from the final graph to the representations of said graph nodes. Likewise, @root@ is a function of the same variety, wrapped in a @Maybe@.
+  (untied, First root) <- fold <$> traverse parseNode (V.toList nodes)
+  -- @untied@ is an intmap, where the keys are graph node IDs and the values are functions from the final graph to the representations of said graph nodes. Likewise, @root@ is a function of the same variety, wrapped in a @Maybe@.
   --
-  -- We define @tied@ as the fixpoint of the former to yield the former as a graph of type @IntMap.IntMap Term@, and apply the latter to said graph to yield the entry point, if any, from which to evaluate.
-  let tied = fix (\ tied -> ($ tied) <$> IntMap.fromList untied)
-  pure (tied, ($ tied) <$> root)
+  -- We define @tied@ as the fixpoint of the former to yield the former as a graph of type @Graph@, and apply the latter to said graph to yield the entry point, if any, from which to evaluate.
+  let tied = fix (\ tied -> ($ Graph tied) <$> untied)
+  pure (Graph tied, ($ Graph tied) <$> root)
 
-parseNode :: A.Object -> A.Parser (IntMap.Key, IntMap.IntMap Term -> Term, Maybe (IntMap.IntMap Term -> Term))
-parseNode o = do
+parseNode :: A.Value -> A.Parser (IntMap.IntMap (Graph -> Term), First (Graph -> Term))
```

Comment:
Can we get a description of what these return parameters are, and how they’re linked to those of `parseGraph`?